include/nanobind/nb_class.h

```diff
@@ -376,6 +376,46 @@ template <typename Arg> struct init_implicit {
     }
 };
 
+template <typename Func, typename Sig = detail::function_signature_t<Func>>
+struct new_;
+
+template <typename Func, typename Return, typename... Args>
+struct new_<Func, Return(Args...)> {
+    std::remove_reference_t<Func> func;
+
+    new_(Func &&f) : func((detail::forward_t<Func>) f) {}
+
+    template <typename Class, typename... Extra>
+    NB_INLINE void execute(Class &cl, const Extra&... extra) {
+        using Type = typename Class::Type;
+        object existing = cl.attr("__new__");
+        if (PyCFunction_Check(existing.ptr()) && sizeof...(Args) != 0) {
+            // This is the original nanobind __new__. If we overwrite it, we'll
+            // have no way to create an object of this type "the normal way",
+            // which would make it difficult to support pickling for this type.
+            // To maximize flexibility here, we provide a cpp_function __new__
+            // (which can be overloaded) to wrap the slot __new__ (which can't).
+            // If the incoming __new__ has no arguments, then this overload
+            // would be ambiguous with the user-supplied one, so we don't add
+            // it; the user will have to figure out something else for pickle.
+            cl.def_static("__new__",
+                          [](handle type) {
+                              if (!type_check(type))
+                                  detail::raise_cast_error();
```

Comment:
perhaps a type error is more appropriate?

Reply:
This is what you hit if you write e.g. `t.UniqueInt.__new__(int)`. I think someone who gets here probably knows what they did wrong; my main priority is that it shouldn't crash the interpreter. Calling `raise_cast_error` produces more compact code than constructing a new exception, so I went with that.